File: mini-jdbc/src/main/java/com/mini/core/jdbc/AbstractDao.java
package com.mini.core.jdbc;

import com.mini.core.jdbc.builder.SQLBuilder;
import com.mini.core.jdbc.model.Page;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.jdbc.core.RowMapper;

import java.sql.Date;
import java.sql.Time;
import java.sql.Timestamp;
import java.util.List;
import java.util.Map;

public abstract class AbstractDao implements JdbcInterface {
    protected abstract JdbcTemplate writeTemplate();

    protected abstract JdbcTemplate readTemplate();

    @Override
    public int execute(@NotNull String sql, Object... params) {
        return writeTemplate().execute(sql, params);
    }

    @Override
    public int execute(@NotNull SQLBuilder builder) {
        return writeTemplate().execute(builder);
    }

    @NotNull
    @Override
    public int[] executeBatch(@NotNull String... sql) {
        return writeTemplate().executeBatch(sql);
    }

    @NotNull
    @Override
    public int[] executeBatch(@NotNull String sql, Object[]... paramsArray) {
        return writeTemplate().executeBatch(sql, paramsArray);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(@NotNull String sql, RowMapper<T> mapper, Object... params) {
        return readTemplate().queryList(sql, mapper, params);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(@NotNull SQLBuilder builder, @NotNull RowMapper<T> mapper) {
        return readTemplate().queryList(builder, mapper);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(@NotNull String sql, @NotNull Class<T> type, Object... params) {
        return readTemplate().queryList(sql, type, params);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(@NotNull SQLBuilder builder, @NotNull Class<T> type) {
        return readTemplate().queryList(builder, type);
    }

    @NotNull
    @Override
    public List<Map<String, Object>> queryListMap(@NotNull String sql, Object... params) {
        return readTemplate().queryListMap(sql, params);
    }

    @NotNull
    @Override
    public List<Map<String, Object>> queryListMap(@NotNull SQLBuilder builder) {
        return readTemplate().queryListMap(builder);
    }

    @NotNull
    @Override
    public <T> List<T> queryListSingle(@NotNull String sql, @NotNull Class<T> type, Object... params) {
        return readTemplate().queryListSingle(sql, type, params);
    }

    @NotNull
    @Override
    public <T> List<T> queryListSingle(@NotNull SQLBuilder builder, Class<T> type) {
        return readTemplate().queryListSingle(builder, type);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(int start, int limit, @NotNull String sql, @NotNull RowMapper<T> mapper, Object... params) {
        return readTemplate().queryList(sql, mapper, params);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(int start, int limit, @NotNull SQLBuilder builder, RowMapper<T> mapper) {
        return readTemplate().queryList(start, limit, builder, mapper);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(int start, int limit, @NotNull String sql, @NotNull Class<T> type, Object... params) {
        return readTemplate().queryList(start, limit, sql, type, params);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(int start, int limit, @NotNull SQLBuilder builder, @NotNull Class<T> type) {
        return readTemplate().queryList(start, limit, builder, type);
    }

    @NotNull
    @Override
    public List<Map<String, Object>> queryListMap(int start, int limit, @NotNull String sql, Object... params) {
        return readTemplate().queryListMap(start, limit, sql, params);
    }

    @NotNull
    @Override
    public List<Map<String, Object>> queryListMap(int start, int limit, @NotNull SQLBuilder builder) {
        return readTemplate().queryListMap(start, limit, builder);
    }

    @NotNull
    @Override
    public <T> List<T> queryListSingle(int start, int limit, @NotNull String sql, @NotNull Class<T> type, Object... params) {
        return readTemplate().queryListSingle(start, limit, sql, type, params);
    }

    @NotNull
    @Override
    public <T> List<T> queryListSingle(int start, int limit, @NotNull SQLBuilder builder, @NotNull Class<T> type) {
        return readTemplate().queryListSingle(start, limit, builder, type);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(int limit, @NotNull String sql, @NotNull RowMapper<T> mapper, Object... params) {
        return readTemplate().queryList(limit, sql, mapper, params);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(int limit, @NotNull SQLBuilder builder, @NotNull RowMapper<T> mapper) {
        return readTemplate().queryList(limit, builder, mapper);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(int limit, String sql, @NotNull Class<T> type, Object... params) {
        return readTemplate().queryList(limit, sql, type, params);
    }

    @NotNull
    @Override
    public <T> List<T> queryList(int limit, @NotNull SQLBuilder builder, @NotNull Class<T> type) {
        return readTemplate().queryList(limit, builder, type);
    }

    @NotNull
    @Override
    public List<Map<String, Object>> queryListMap(int limit, @NotNull String sql, Object... params) {
        return readTemplate().queryListMap(limit, sql, params);
    }

    @NotNull
    @Override
    public List<Map<String, Object>> queryListMap(int limit, @NotNull SQLBuilder builder) {
        return readTemplate().queryListMap(limit, builder);
    }

    @NotNull
    @Override
    public <T> List<T> queryListSingle(int limit, @NotNull String sql, @NotNull Class<T> type, Object... params) {
        return readTemplate().queryListSingle(limit, sql, type, params);
    }

    @NotNull
    @Override
    public <T> List<T> queryListSingle(int limit, @NotNull SQLBuilder builder, @NotNull Class<T> type) {
        return readTemplate().queryListSingle(limit, builder, type);
    }

    @Nullable
    @Override
    public <T> T queryObject(@NotNull String sql, @NotNull RowMapper<T> mapper, Object... params) {
        return readTemplate().queryObject(sql, mapper, params);
    }

    @Nullable
    @Override
    public <T> T queryObject(@NotNull SQLBuilder builder, @NotNull RowMapper<T> mapper) {
        return readTemplate().queryObject(builder, mapper);
    }

    @Nullable
    @Override
    public <T> T queryObject(@NotNull String sql, @NotNull Class<T> type, Object... params) {
        return readTemplate().queryObject(sql, type, params);
    }

    @Nullable
    @Override
    public <T> T queryObject(@NotNull SQLBuilder builder, @NotNull Class<T> type) {
        return readTemplate().queryObject(builder, type);
    }

    @Nullable
    @Override
    public Map<String, Object> queryObjectMap(@NotNull String sql, Object... params) {
        return readTemplate().queryObjectMap(sql, params);
    }

    @Nullable
    @Override
    public Map<String, Object> queryObjectMap(@NotNull SQLBuilder builder) {
        return readTemplate().queryObjectMap(builder);
    }

    @Nullable
    @Override
    public <T> T queryObjectSingle(@NotNull String sql, @NotNull Class<T> type, Object... params) {
        return readTemplate().queryObjectSingle(sql, type, params);
    }

    @Nullable
    @Override
    public <T> T queryObjectSingle(@NotNull SQLBuilder builder, @NotNull Class<T> type) {
        return readTemplate().queryObjectSingle(builder, type);
    }

    @Nullable
    @Override
    public String queryString(String sql, Object... params) {
        return readTemplate().queryString(sql, params);
    }

    @Nullable
    @Override
    public String queryString(@NotNull SQLBuilder builder) {
        return readTemplate().queryString(builder);
    }

    @Nullable
    @Override
    public Long queryLong(@NotNull String sql, Object... params) {
        return readTemplate().queryLong(sql, params);
    }

    @Nullable
    @Override
    public Long queryLong(@NotNull SQLBuilder builder) {
        return readTemplate().queryLong(builder);
    }

    @Nullable
    @Override
    public Integer queryInt(@NotNull String sql, Object... params) {
        return readTemplate().queryInt(sql, params);
    }

    @Nullable
    @Override
    public Integer queryInt(@NotNull SQLBuilder builder) {
        return readTemplate().queryInt(builder);
    }

    @Nullable
    @Override
    public Short queryShort(@NotNull String sql, Object... params) {
        return readTemplate().queryShort(sql, params);
    }

    @Nullable
    @Override
    public Short queryShort(@NotNull SQLBuilder builder) {
        return readTemplate().queryShort(builder);
    }

    @Nullable
    @Override
    public Byte queryByte(@NotNull String sql, Object... params) {
        return readTemplate().queryByte(sql, params);
    }

    @Nullable
    @Override
    public Byte queryByte(@NotNull SQLBuilder builder) {
        return readTemplate().queryByte(builder);
    }

    @Nullable
    @Override
    public Double queryDouble(@NotNull String sql, Object... params) {
        return readTemplate().queryDouble(sql, params);
    }

    @Nullable
    @Override
    public Double queryDouble(@NotNull SQLBuilder builder) {
        return readTemplate().queryDouble(builder);
    }

    @Nullable
    @Override
    public Float queryFloat(@NotNull String sql, Object... params) {
        return readTemplate().queryFloat(sql, params);
    }

    @Nullable
    @Override
    public Float queryFloat(@NotNull SQLBuilder builder) {
        return readTemplate().queryFloat(builder);
    }

    @Nullable
    @Override
    public Boolean queryBoolean(@NotNull String sql, Object... params) {
        return readTemplate().queryBoolean(sql, params);
    }

    @Nullable
    @Override
    public Boolean queryBoolean(@NotNull SQLBuilder builder) {
        return readTemplate().queryBoolean(builder);
    }

    @Nullable
    @Override
    public Timestamp queryTimestamp(@NotNull String sql, Object... params) {
        return readTemplate().queryTimestamp(sql, params);
    }

    @Nullable
    @Override
    public Timestamp queryTimestamp(@NotNull SQLBuilder builder) {
        return readTemplate().queryTimestamp(builder);
    }

    @Nullable
    @Override
    public Date queryDate(@NotNull String sql, Object... params) {
        return readTemplate().queryDate(sql, params);
    }

    @Nullable
    @Override
    public Date queryDate(@NotNull SQLBuilder builder) {
        return readTemplate().queryDate(builder);
    }

    @Nullable
    @Override
    public Time queryTime(@NotNull String sql, Object... params) {
        return readTemplate().queryTime(sql, params);
    }

    @Nullable
    @Override
    public Time queryTime(@NotNull SQLBuilder builder) {
        return readTemplate().queryTime(builder);
    }

    @NotNull
    @Override
    public <T> Page<T> queryPage(int page, int limit, @NotNull String sql, @NotNull RowMapper<T> mapper, Object... params) {
        return readTemplate().queryPage(page, limit, sql, mapper, params);
    }

    @NotNull
    @Override
    public <T> Page<T> queryPage(int page, int limit, @NotNull SQLBuilder builder, @NotNull RowMapper<T> mapper) {
        return readTemplate().queryPage(page, limit, builder, mapper);
    }

    @NotNull
    @Override
    public <T> Page<T> queryPage(int page, int limit, String sql, @NotNull Class<T> type, Object... params) {
        return readTemplate().queryPage(page, limit, sql, type, params);
    }

    @NotNull
    @Override
    public <T> Page<T> queryPage(int page, int limit, @NotNull SQLBuilder builder, Class<T> type) {
        return readTemplate().queryPage(page, limit, builder, type);
    }

    @NotNull
    @Override
    public Page<Map<String, Object>> queryPageMap(int page, int limit, @NotNull String sql, Object... params) {
        return readTemplate().queryPageMap(page, limit, sql, params);
    }

    @NotNull
    @Override
    public Page<Map<String, Object>> queryPageMap(int page, int limit, @NotNull SQLBuilder builder) {
        return readTemplate().queryPageMap(page, limit, builder);
    }

    @NotNull
    @Override
    public <T> Page<T> queryPageSingle(int page, int limit, @NotNull String sql, @NotNull Class<T> type, Object... params) {
        return readTemplate().queryPageSingle(page, limit, sql, type, params);
    }

    @NotNull
    @Override
    public <T> Page<T> queryPageSingle(int page, int limit, @NotNull SQLBuilder builder, @NotNull Class<T> type) {
        return readTemplate().queryPageSingle(page, limit, builder, type);
    }
}

File: mini-jdbc/src/main/kotlin/com/mini/core/jdbc/AbstractDao.kt
@file:Suppress("unused")

package com.mini.core.jdbc

import com.mini.core.jdbc.builder.SQLBuilder
import com.mini.core.jdbc.model.Paging
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.support.GeneratedKeyHolder
import java.sql.Date
import java.sql.Time
import java.sql.Timestamp
import kotlin.reflect.KClass

abstract class AbstractDao : JdbcInterface {
    internal abstract fun writeTemplate(): JdbcTemplate
    internal abstract fun readTemplate(): JdbcTemplate

    final override fun execute(sql: String, vararg params: Any?): Int {
        return writeTemplate().execute(sql, *params)
    }

    final override fun execute(builder: SQLBuilder): Int {
        return writeTemplate().execute(builder)
    }

    final override fun execute(sql: String, holder: GeneratedKeyHolder, vararg params: Any?): Int {
        return writeTemplate().execute(sql, holder, *params)
    }

    final override fun execute(builder: SQLBuilder, holder: GeneratedKeyHolder): Int {
        return writeTemplate().execute(builder, holder)
    }

    final override fun executeBatch(vararg sql: String): IntArray {
        return writeTemplate().executeBatch(*sql)
    }

    final override fun executeBatch(sql: String, batchArgs: List<Array<Any>>): IntArray {
        return writeTemplate().executeBatch(sql, batchArgs)
    }

    final override fun executeBatch(sql: String, batchArgs: List<Array<Any>>, argTypes: IntArray): IntArray {
        return writeTemplate().executeBatch(sql, batchArgs, argTypes)
    }

    final override fun <T : Any> queryList(sql: String, m: RowMapper<out T?>, vararg params: Any?): List<T?> {
        return readTemplate().queryList(sql, m, *params)
    }

    final override fun <T : Any> queryList(builder: SQLBuilder, m: RowMapper<out T?>): List<T?> {
        return readTemplate().queryList(builder, m)
    }

    final override fun <T : Any> queryList(sql: String, type: KClass<out T>, vararg params: Any?): List<T?> {
        return readTemplate().queryList(sql, type, *params)
    }

    final override fun <T : Any> queryList(builder: SQLBuilder, type: KClass<out T>): List<T?> {
        return readTemplate().queryList(builder, type)
    }


    final override fun <T : Any> queryListSingle(sql: String, type: KClass<out T>, vararg params: Any?): List<T?> {
        return readTemplate().queryListSingle(sql, type, *params)
    }


    final override fun <T : Any> queryListSingle(builder: SQLBuilder, type: KClass<out T>): List<T?> {
        return queryListSingle(builder.SQL, type, builder.ARGS)
    }

    final override fun <T : Any> queryObject(sql: String, m: RowMapper<out T?>, vararg params: Any?): T? {
        return readTemplate().queryObject(sql, m, *params)
    }

    final override fun <T : Any> queryObject(builder: SQLBuilder, m: RowMapper<out T?>): T? {
        return readTemplate().queryObject(builder, m)
    }

    final override fun <T : Any> queryObject(sql: String, type: KClass<out T>, vararg params: Any?): T? {
        return readTemplate().queryObject(sql, type, *params)
    }

    final override fun <T : Any> queryObject(builder: SQLBuilder, type: KClass<out T>): T? {
        return readTemplate().queryObject(builder, type)
    }

    final override fun <T : Any> queryObjectSingle(sql: String, type: KClass<out T>, vararg params: Any?): T? {
        return readTemplate().queryObjectSingle(sql, type, *params)
    }

    final override fun <T : Any> queryObjectSingle(builder: SQLBuilder, type: KClass<out T>): T? {
        return readTemplate().queryObjectSingle(builder, type)
    }

    final override fun <T : Any> queryList(start: Int, limit: Int, sql: String, m: RowMapper<out T?>, vararg params: Any?): List<T?> {
        return readTemplate().queryList(start, limit, sql, m, *params)
    }

    final override fun <T : Any> queryList(start: Int, limit: Int, builder: SQLBuilder, m: RowMapper<out T?>): List<T?> {
        return readTemplate().queryList(start, limit, builder, m)
    }

    final override fun <T : Any> queryList(start: Int, limit: Int, sql: String, type: KClass<out T>, vararg params: Any?): List<T?> {
        return readTemplate().queryList(start, limit, sql, type, *params)
    }

    final override fun <T : Any> queryList(start: Int, limit: Int, builder: SQLBuilder, type: KClass<out T>): List<T?> {
        return readTemplate().queryList(start, limit, builder, type)
    }


    final override fun <T : Any> queryListSingle(start: Int, limit: Int, sql: String, type: KClass<out T>, vararg params: Any?): List<T?> {
        return readTemplate().queryListSingle(start, limit, sql, type, *params)
    }


    final override fun <T : Any> queryListSingle(start: Int, limit: Int, builder: SQLBuilder, type: KClass<out T>): List<T?> {
        return readTemplate().queryListSingle(start, limit, builder, type)
    }

    final override fun <T : Any> queryList(limit: Int, sql: String, m: RowMapper<out T?>, vararg params: Any?): List<T?> {
        return readTemplate().queryList(limit, sql, m, *params)
    }

    final override fun <T : Any> queryList(limit: Int, builder: SQLBuilder, m: RowMapper<out T?>): List<T?> {
        return readTemplate().queryList(limit, builder, m)
    }

    final override fun <T : Any> queryList(limit: Int, sql: String, type: KClass<out T>, vararg params: Any?): List<T?> {
        return readTemplate().queryList(limit, sql, type, *params)
    }

    final override fun <T : Any> queryList(limit: Int, builder: SQLBuilder, type: KClass<out T>): List<T?> {
        return readTemplate().queryList(limit, builder, type)
    }

    final override fun <T : Any> queryListSingle(limit: Int, sql: String, type: KClass<out T>, vararg params: Any?): List<T?> {
        return readTemplate().queryListSingle(limit, sql, type, *params)
    }

    final override fun <T : Any> queryListSingle(limit: Int, builder: SQLBuilder, type: KClass<out T>): List<T?> {
        return readTemplate().queryListSingle(limit, builder, type)
    }

    final override fun queryString(sql: String, vararg params: Any?): String? {
        return readTemplate().queryString(sql, *params)
    }

    final override fun queryString(builder: SQLBuilder): String? {
        return readTemplate().queryString(builder)
    }

    final override fun queryLong(sql: String, vararg params: Any?): Long? {
        return readTemplate().queryLong(sql, *params)
    }

    final override fun queryLong(builder: SQLBuilder): Long? {
        return readTemplate().queryLong(builder)
    }

    final override fun queryInt(sql: String, vararg params: Any?): Int? {
        return readTemplate().queryInt(sql, *params)
    }

    final override fun queryInt(builder: SQLBuilder): Int? {
        return readTemplate().queryInt(builder)
    }

    final override fun queryShort(sql: String, vararg params: Any?): Short? {
        return readTemplate().queryShort(sql, *params)
    }

    final override fun queryShort(builder: SQLBuilder): Short? {
        return readTemplate().queryShort(builder)
    }

    final override fun queryByte(sql: String, vararg params: Any?): Byte? {
        return readTemplate().queryByte(sql, *params)
    }

    final override fun queryByte(builder: SQLBuilder): Byte? {
        return readTemplate().queryByte(builder)
    }

    final override fun queryDouble(sql: String, vararg params: Any?): Double? {
        return readTemplate().queryDouble(sql, *params)
    }

    final override fun queryDouble(builder: SQLBuilder): Double? {
        return readTemplate().queryDouble(builder)
    }

    final override fun queryFloat(sql: String, vararg params: Any?): Float? {
        return readTemplate().queryFloat(sql, *params)
    }

    final override fun queryFloat(builder: SQLBuilder): Float? {
        return readTemplate().queryFloat(builder)
    }

    final override fun queryBoolean(sql: String, vararg params: Any?): Boolean? {
        return readTemplate().queryBoolean(sql, *params)
    }

    final override fun queryBoolean(builder: SQLBuilder): Boolean? {
        return readTemplate().queryBoolean(builder)
    }

    final override fun queryTimestamp(sql: String, vararg params: Any?): Timestamp? {
        return readTemplate().queryTimestamp(sql, *params)
    }

    final override fun queryTimestamp(builder: SQLBuilder): Timestamp? {
        return readTemplate().queryTimestamp(builder)
    }

    final override fun queryDate(sql: String, vararg params: Any?): Date? {
        return readTemplate().queryDate(sql, *params)
    }

    final override fun queryDate(builder: SQLBuilder): Date? {
        return readTemplate().queryDate(builder)
    }

    final override fun queryTime(sql: String, vararg params: Any?): Time? {
        return readTemplate().queryTime(sql, *params)
    }

    final override fun queryTime(builder: SQLBuilder): Time? {
        return readTemplate().queryTime(builder)
    }

    final override fun <T : Any> queryPaging(page: Int, limit: Int, sql: String, m: RowMapper<out T?>, vararg params: Any?): Paging<T?> {
        return readTemplate().queryPaging(page, limit, sql, m, *params)
    }

    final override fun <T : Any> queryPaging(page: Int, limit: Int, builder: SQLBuilder, m: RowMapper<out T?>): Paging<T?> {
        return readTemplate().queryPaging(page, limit, builder, m)
    }

    final override fun <T : Any> queryPaging(page: Int, limit: Int, sql: String, type: KClass<out T>, vararg params: Any?): Paging<T?> {
        return readTemplate().queryPaging(page, limit, sql, type, *params)
    }

    final override fun <T : Any> queryPaging(page: Int, limit: Int, builder: SQLBuilder, type: KClass<out T>): Paging<T?> {
        return readTemplate().queryPaging(page, limit, builder, type)
    }

    final override fun <T : Any> queryPagingSingle(page: Int, limit: Int, sql: String, type: KClass<out T>, vararg params: Any?): Paging<T?> {
        return readTemplate().queryPagingSingle(page, limit, sql, type, *params)
    }

    final override fun <T : Any> queryPagingSingle(page: Int, limit: Int, builder: SQLBuilder, type: KClass<out T>): Paging<T?> {
        return readTemplate().queryPagingSingle(page, limit, builder, type)
    }
}